### ISSUE TITLE
How to run distributed TensorFlow

### ISSUE THREAD

**Issue by wangshuaizs** [bytedance/byteps#26]
Hi,

I want to run distributed tensorflow in my physical machine, though the scheduler, the server and the only worker are in the same physical machine (without docker). I only installed tensorflow, without mxnet or pytorch. When I run cmd 
```
$ DMLC_ROLE=scheduler DMLC_PS_ROOT_URI=12.12.10.12 DMLC_PS_ROOT_PORT=9000 DMLC_NUM_WORKER=1 DMLC_NUM_SERVER=1 python launcher/launch.py
```
the output is 
```
BytePS launching scheduler
BYTEPS_SERVER_MXNET_PATH env not set
```
And the server's output is similar. Do I have to install mxnet to run distributed training, though I only use tensorflow?

**Comment by wangshuaizs** [comment]
Also, I found that if I only start worker with the following cmd :
```
$ DMLC_ROLE=worker DMLC_PS_ROOT_URI=12.12.10.12 DMLC_PS_ROOT_PORT=9000 DMLC_WORKER_ID=0 DMLC_NUM_WORKER=1 DMLC_NUM_SERVER=1 python launcher/launch.py python example/tensorflow/synthetic_benchmark.py
``` 
the worker can run immediately before I start server. So, I wonder if the worker communicate with the server?

**Comment by bobzhuyb** [comment]
When DMLC_NUM_WORKER is 1, BytePS will not use PS at all because it is not needed. However, if you really want to test this case, you can set BYTEPS_FORCE_DISTRIBUTED=1 as described in https://github.com/bytedance/byteps/blob/master/docs/env.md

**Comment by bobzhuyb** [comment]
"BYTEPS_SERVER_MXNET_PATH env not set" is indeed a problem. Are you using our official docker image? If not, you have to compile MXNet-based server (see README.md). We highly recommend you to just use our docker image for the server and scheduler. The command to start the docker can be found in step-by-step tutorial.

**Comment by ymjiang** [comment]
@wangshuaizs As described in README.md:
"For your server and scheduler node, we highly recommend you to just use our prebuilt docker image bytepsimage/byteps_server. Otherwise, you have to manually compile our modified MXNet as in our Dockerfile."

if you don't use our docker image as the server then you will have to compile the server by yourself (yes you need to compile MXNet from scratch, though your worker is just using Tensorflow).

**Comment by wangshuaizs** [comment]
@bobzhuyb @ymjiang , thanks. After installing mxnet, it worked. 

But based on my experiment with 2 physical servers, each with one k40 GPU. The link between two servers is 40G TCP. Each server has one worker, and parameter server and scheduler run in one of them. The total training speed of two GPUs is only 73.8 imgs/sec, while the training speed of single GPU is 53.1 imgs/sec. Is the performance normal for byteps? It seems that it is worse than the traditional parameter server.

**Comment by ymjiang** [comment]
@wangshuaizs Can you provide the running scripts (the model you use, and the batch size, etc)? What is the expected performance, given that you claim the performance is worse than traditional PS?

Note: You can try to increase the number of PS instances, i.e., multiple PS on one physical machines. 
Also, here are some performance tuning tips: https://github.com/bytedance/byteps/blob/master/docs/env.md#performance-tuning

**Comment by bobzhuyb** [comment]
@wangshuaizs You need to run the parameter servers in different physical servers from the workers, and the number of parameters servers should be >= the number of workers. Otherwise, it would be roughly the same as allreduce.

I will add a best practice documentation today.

By the way, you said you installed mxnet. Did you build it from our MXNet repo https://github.com/bytedance/incubator-mxnet, or you just installed the official one? The official MXNet performance will be much worse.

**Comment by ymjiang** [comment]
Actually I notice your setup is setting server and worker together on one physical machine. This does not meet BytePS requirement, as explained [here](https://github.com/bytedance/byteps/blob/master/docs/rationale.md). You should put them apart.

**Comment by wangshuaizs** [comment]
@ymjiang , the scripts I used is the one in examples ```https://github.com/bytedance/byteps/blob/cb6913d056c467e9e655a5f7c6b72edb583ef863/example/tensorflow/synthetic_benchmark.py```, with default batch size and etc.

@bobzhuyb Yes, I built MXNet using your MXNet repo and your install cmds shown in ```https://github.com/bytedance/byteps/blob/master/docker/Dockerfile.server```. And I will try to run ps in different machines.

**Comment by bobzhuyb** [comment]
@wangshuaizs  See this, https://github.com/bytedance/byteps/blob/master/docs/best-practice.md
The distributed mode part.

**Comment by wangshuaizs** [comment]
@bobzhuyb, thanks, this documentation is very clear. I will close this issue.
